pkg/admission/cert/writer/fs.go

```diff
@@ -16,125 +16,175 @@ limitations under the License.
+		if strings.HasPrefix(k, FSCertProvisionAnnotationKeyPrefix) {
+			webhookName := strings.TrimPrefix(k, FSCertProvisionAnnotationKeyPrefix)
+			webhookMap[webhookName] = &webhookAndPath{
+				path: v,
+			}
+		}
+	}
+
+	webhooks, err := getWebhooksFromObject(webhookConfig)
+	if err != nil {
+		return err
+	}
+	for i, webhook := range webhooks {
+		if p, found := webhookMap[webhook.Name]; found {
+			p.webhook = &webhooks[i]
+		}
+	}
+
+	// validation
+	for k, v := range webhookMap {
+		if v.webhook == nil {
+			return fmt.Errorf("expecting a webhook named %q", k)
+		}
+	}
+
+	generator := s.CertGenerator
+	if s.CertGenerator == nil {
+		generator = &certgenerator.SelfSignedCertGenerator{}
+	}
+
+	cw := &fsCertWriter{
+		certGenerator: generator,
+		webhookConfig: webhookConfig,
+		webhookMap:    webhookMap,
+	}
+	return cw.ensureCert()
+}
+
+// fsCertWriter deals with writing to the local filesystem.
+type fsCertWriter struct {
+	certGenerator certgenerator.CertGenerator
+
+	webhookConfig runtime.Object
+	webhookMap    map[string]*webhookAndPath
+}
+
+type webhookAndPath struct {
+	webhook *admissionregistrationv1beta1.Webhook
+	path    string
+}
+
+var _ certReadWriter = &fsCertWriter{}
+
+func (s *fsCertWriter) ensureCert() error {
+	var err error
+	for _, v := range s.webhookMap {
+		err = handleCommon(v.webhook, s)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *fsCertWriter) write(webhookName string) (*certgenerator.Artifacts, error) {
+	return f.doWrite(webhookName)
+}
+
+func (f *fsCertWriter) overwrite(webhookName string) (*certgenerator.Artifacts, error) {
+	return f.doWrite(webhookName)
+}
+
+func (f *fsCertWriter) doWrite(webhookName string) (*certgenerator.Artifacts, error) {
+	v := f.webhookMap[webhookName]
+	commonName, err := dnsNameForWebhook(&v.webhook.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := f.certGenerator.Generate(commonName)
+	if err != nil {
+		return nil, err
+	}
+	aw, err := atomic.NewAtomicWriter(v.path, fmt.Sprintf("processing webhook %q", webhookName))
+	if err != nil {
+		return nil, err
+	}
+	err = aw.Write(certToProjectionMap(certs))
+	return certs, err
+}
+
+func (f *fsCertWriter) read(webhookName string) (*certgenerator.Artifacts, error) {
+	dir := f.webhookMap[webhookName].path
+	caBytes, err := ioutil.ReadFile(path.Join(dir, CACertName))
+	if err != nil {
+		return nil, err
+	}
+	certBytes, err := ioutil.ReadFile(path.Join(dir, ServerCertName))
+	if err != nil {
+		return nil, err
+	}
+	keyBytes, err := ioutil.ReadFile(path.Join(dir, ServerKeyName))
+	if err != nil {
+		return nil, err
+	}
+	return &certgenerator.Artifacts{
+		CACert: caBytes,
+		Cert:   certBytes,
+		Key:    keyBytes,
+	}, nil
+}
+
+func certToProjectionMap(cert *certgenerator.Artifacts) map[string]atomic.FileProjection {
+	return map[string]atomic.FileProjection{
+		CACertName: {
+			Data: cert.CACert,
+			Mode: 0666,
```

Comment:
I'm curious about this mode. Does it really need to be open to everyone?

Reply:
Added a TODO.
Let's figure it out later :)